### ISSUE TITLE
Help us build challenges!

### ISSUE THREAD

**Issue by merwanehamadi** [Significant-Gravitas/Auto-GPT#3835]
### Summary 💡

Challenges are tasks Auto-GPT is not able to achieve. Challenges will help us improve Auto-GPT

We need your help to build these challenges and the ecosystem around them

Here is a breakdown of the help we need.

# A-Challenge Creation
### Memory
- ~~https://github.com/Significant-Gravitas/Auto-GPT/issues/3838 => @dschonholtz~~
- Memory challenge D => @Androbin
### Information Retrieval
- ~https://github.com/Significant-Gravitas/Auto-GPT/issues/3837 => @PortlandKyGuy~
- Information Retrieval Challenge C => @Androbin
### Research
- Research Challenge A => @Androbin
### Psychological
- ~https://github.com/Significant-Gravitas/Auto-GPT/issues/3871 => unassigned~
### Debug Code
- ~https://github.com/Significant-Gravitas/Auto-GPT/issues/3836 @gravelBridge~
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3935 @gravelBridge

### Adaptability
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3937 => unassigned
### Website Navigation Challenge
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3936 => unassigned

### Self Improvement (Solve challenges automatically)
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3912 => unassigned

### Automated Challenge Creation 
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3917 => unassigned

### Basic Abilities 
- ~https://github.com/Significant-Gravitas/Auto-GPT/issues/4233 create a BIG FILE challenged => unassigned~

## 2-Design brand new challenge categories
- Alignment
- Obtain Knowledge
- Focus (stay on task)
- Planning (suggested by @Boostrix)
- Others if you have ideas.
DM me if interested (discord below)

# Challenges Auto-GPT can already perform
We currently have what we call regression tests. You can imagine regression tests as challenges Auto-GPT is already able to perform.
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3901 => unassigned

# UX experience around challenges
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3839 => unassigned

# Improve logs/DEBUG folder to help people beat challenges more easily
The logs/DEBUG folder allows everyone to understand what Auto-GPT is doing at every cycle
![Screenshot 2023-05-05 at 7 32 57 AM](https://user-images.githubusercontent.com/9652976/236492792-548868e8-7ebc-49ee-92de-e6a6d23153e3.png)
We need to:
- ~~https://github.com/Significant-Gravitas/Auto-GPT/issues/3842 => @AndresCdo  done thank you !~~ 
- ~~https://github.com/Significant-Gravitas/Auto-GPT/issues/3843 => @AndresCdo  done thank you !~~
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3960 => log ai_settings.yaml
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3844 => BLOCKED =>waiting for @Pwuts memory rework to assign
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3957 => BLOCKED => waiting for rearch

# "Fix Auto-GPT" challenges !
The vision of the fix auto-gpt challenges is to give the tools for the community to report wrong behaviors and create challenges around them.
We need to:
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3841 => unassigned
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3847 => unassigned

# Make it easy for people to change the prompt
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3858 => @Eggrolling-hu
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3954 => idea brought up by @Boostrix 

# Build the CI pipeline !
DM me on discord if you have devops experience and want to help us build the pipeline that will allow people to submit their challenges and see how they perform!
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3966 => unassigned

# Pytest Imrovements
- ~~https://github.com/Significant-Gravitas/Auto-GPT/issues/3863 => @AndresCdo~~
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3900 => unassigned
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3906 => unassigned
- https://github.com/Significant-Gravitas/Auto-GPT/issues/4199 => unassigned
- https://github.com/Significant-Gravitas/Auto-GPT/issues/4496 => unassigned
# Challenges refactorization
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3883 => unassigned
- https://github.com/Significant-Gravitas/Auto-GPT/issues/3907 => unassigned
- ~~https://github.com/Significant-Gravitas/Auto-GPT/issues/4161 => @merwanehamadi~~
- ~~https://github.com/Significant-Gravitas/Auto-GPT/issues/4189 => unassigned~~
- https://github.com/Significant-Gravitas/Auto-GPT/issues/4190 => unassigned
# CI pipeline improvements
- ~https://github.com/Significant-Gravitas/Auto-GPT/issues/4163 => canceled~
- ~https://github.com/Significant-Gravitas/Auto-GPT/issues/4196 => @merwanehamadi~
-~~allow ci pipeline to make calls to api providers => @merwanehamadi~~
- ~https://github.com/Significant-Gravitas/Auto-GPT/issues/4461 => @erik-megarad~
- ~https://github.com/Significant-Gravitas/Auto-GPT/issues/4258 => @merwanehamadi~
-https://github.com/Significant-Gravitas/Auto-GPT/issues/4590
#####################
(issue to create: 
- create a test that makes sure the challenges have a corresponding doc
- create a test maker, i.e a command that creates a test automatically
)

Discord (merwanehamadi)
Join Auto-GPT's discord channel: https://discord.gg/autogpt

**Comment by Boostrix** [comment]
The section titled "others": should at least mention "planning", which is severely lacking at the moment. I bet all of us can come up with dozens of ai_settings files (objectives) where it fails to plan properly, or where it fails to recognize dependencies in between tasks, but also recognizing that it previously succeeded/completed a task and should proceed: https://github.com/Significant-Gravitas/Auto-GPT/issues/3593#issuecomment-1532870949

**Comment by merwanehamadi** [comment]
Yeah this is a great suggestion @Boostrix can I talk to you in discord ? my discord is merwanehamadi.
https://discord.gg/autogpt

**Comment by anonhostpi** [comment]
Boostrix has preference for staying off of Discord based off my prior interactions with them

**Comment by merwanehamadi** [comment]
@Boostrix can you think of a challenge we could build in order to test planning skills ?

**Comment by merwanehamadi** [comment]
@Androbin has suggested a very nice memory challenge involving files that are read in the wrong order. More details coming soon hopefully.

**Comment by Boostrix** [comment]
> can you think of a challenge we could build in order to test planning skills ?

For starters, I would consider a plan to be multi-objective task with multiple dimensions of leeway and constraints that the agent needs to explore.

So I suppose anything involving detangling dependencies (or lack thereof) should work to get this started. That would involve organizing steps but also coordinating in between steps. In #3593, @dschonholtz is juggling some nice ideas. And given the current state of things, it might actually be good to have working/non-working examples of plans for the agent, so that we can see which approach(es) look promising. 

And in fact, GPT itself seems pretty good at coming up with potential candidates:

- Grocery shopping: An AI agent could plan a grocery shopping trip by identifying the necessary items, determining the optimal route to take through the store to minimize time and energy, and considering any budget or dietary constraints.
- Booking a flight: An AI agent could plan a flight booking by identifying the preferred travel dates, considering any budget or scheduling constraints, and selecting the best flight options based on price, duration, and other criteria.
- Event planning: An AI agent could plan an event by identifying the necessary resources (such as venue, catering, and decorations), determining the optimal timing and scheduling, and considering any budget or other constraints.
- Software development: An AI agent could plan a software development project by identifying the necessary features and functionality, breaking them down into smaller tasks and subtasks, and assigning them to team members based on their skills and availability.
- Construction project management: An AI agent could plan a construction project by identifying the necessary resources (such as labor, materials, and equipment), coordinating with contractors and subcontractors, scheduling tasks and milestones, and monitoring progress to ensure that the project stays on track and within budget.

**Comment by merwanehamadi** [comment]
@Boostrix great suggestions! How do you measure success in a very deterministic way for these items ?
I like the logging resource utilization suggestion. I will create an issue.

The example of regression you gave (entire website) is great! I guess we could use selenium to test the website. It looks like a great project.

**Comment by Boostrix** [comment]
> How do you measure success in a very deterministic way for these items ?
right, good question.

I was thinking to use a simple test case for starters, one where we ask the "planner" to coordinate a meeting between N different people (say 3-4) who are constrained by availability (date/time) - later on, N can be increased, with more constraints added, such as certain people never being available on the same date.
We can then use such a spec to generate a Pyton unit test using GPT to answer for each participant if that participant is available at the requested day/time.

So, the input for the unit test will be date/time for now (for each participant). 
The next step would be to ask Agent-GPT to plan the corresponding meeting by telling it about those constraints and specifying a potential time window (or list thereof).

To verify if the the solution is valid, we merely need to execute the unit test for each participant which will tell us if that participant is available - that way, we're reducing the problem to running tests.

Once we have that fleshed out/working for 3-4 participants, it would make sense to add more complexity to it by adding more options and constraints, including dependencies between these options and constraints.

We could then adapt this framework for other more complex examples (see above).   

> I like the logging resource utilization suggestion. I will create an issue.
For API costs/token (budget) we have several open PRs, number of steps taken is part of at least 2 PRs that I am aware of.
The rest is probably in the realm of #3466 
 
> The example of regression you gave (entire website) is great! I guess we could use selenium to test the website. It looks like a great project.

I didn't even think about using Selenium, I was thinking of treating the final HTML like any XHTML/XML document and query it from Python to see if it's got the relevant tags and attributes as requested by the specs. Personally, I find the whole Selenium stuff mainly useful for highly dynamic websites, static HTML can probably be queried "as is" from Python - no need for Selenium ?

That being said, we could also use @adam-patersons example and add an outer agent to mutate his ai_settings / project.txt file to tinker with 3-5 variations of each step (file names, technology stack, functionality etc) - that way, you end up with a ton of regression tests at the mere cost of a few nested loops.The "deliverables" portion of his specs is such succinct that we could probably use it "as is" to create a corresponding python unit test via Auto-GPT (I have used this for several toy projects, and it's working nicely).

**Comment by Androbin** [comment]
@Boostrix I would agree that our current prompting artificially limits GPT-4 abilities. The issue I see is that we actively discourage long-form chain-of-thought reasoning.

**Comment by Boostrix** [comment]
The idea to use dynamic prompting sounds rather promising:

- #3937

**Comment by Boostrix** [comment]
Here's another good description by @hashratez of an intra-llm-regression that should be suitable to benchmark the system against GPT itself:

#1647
> I then tried the simplest task: create an index.html file with the word "hello". In direct ChatGPT it would output this in 1 second. This took over 10 steps as it writes the code, checks it then does something else, etc.

**Comment by Boostrix** [comment]
we should update the list of potential challenges to add a new category for "experiential tests" - i.e. tests where the agent should be able to apply its experience of previously completing a task, but fails to do so.

The most straightforward example I can think of is it editing/updating a file successfully and 5 minutes later it wants to use interactive editors like nano, vim etc - that's a recurring topic here.

So we should add a benchmark to see how good an agent is at applying experience to future tasks.

A simple test case would be telling it to update a file with some hand holding, and afterwards leaving out the hand holding and counting how many times it succeeds or not (3/10 attempts).

Being able to apply past experience is an essential part of learning.

**Comment by merwanehamadi** [comment]
CI pipeline 4 times faster thanks to @AndresCdo and parallelized tests !

**Comment by Boostrix** [comment]
Suggestion for new challenge type:
In the light of the latest google/DDG dilemma (#4120), we need to harden our commands/agent accordingly (#4157):

- work out a way to randomly make commands malfunction/disabled via pytest (using a timer or a rand function at the command mgr/registry level)
- let the agent figure out that a command is  malfunctioning, so that it comes up with an alternative

This may involve tinkering with different variable/argument substitutions: https://github.com/Significant-Gravitas/Auto-GPT/issues/3904#issuecomment-1545835975

Basically, we need to keep track of commands that previously worked/failed and types of arguments/params that are known to work - including some optional timeout option to retry a command once in a while.

Also, commands really should get access to the error/exception string - because at that point, the LLM can actually help us!

**Comment by merwanehamadi** [comment]
https://github.com/Significant-Gravitas/Auto-GPT/issues/4161  new issue

**Comment by merwanehamadi** [comment]
@Boostrix yeah here you're more in the solution space.
Most of our challenges already require the solution you're describing to be implemented.

**Comment by merwanehamadi** [comment]
@Boostrix could you create an issue for the 2 challenges you mentioned above and label it challenges ?

**Comment by Boostrix** [comment]
Which ones exactly ?

Coming up with a challenge where we mutate a URL should be a pretty straightforward way for the agent tofail, which it still can fix - using URL validation/patching. Most obvious example would be adding whitespaces into the URL without escaping.

**Comment by merwanehamadi** [comment]
all the ones you suggest, we don't have next step for them, can I just have the link of the issues where the challenge idea is written so I put them in this epic ?

**Comment by merwanehamadi** [comment]
New CI pipeline ready: now you can test challenges by creating a Pull Request.
currently working on this so you can SEE the results of the challenges.
https://github.com/Significant-Gravitas/Auto-GPT/issues/4190

**Comment by merwanehamadi** [comment]
thanks @AndresCdo for https://github.com/Significant-Gravitas/Auto-GPT/pull/3868

**Comment by merwanehamadi** [comment]
thank you @PortlandKyGuy for your work! https://github.com/Significant-Gravitas/Auto-GPT/pull/4261

**Comment by merwanehamadi** [comment]
thank you @erik-megarad https://github.com/Significant-Gravitas/Auto-GPT/pull/4469

**Comment by merwanehamadi** [comment]
thank you @javableu  !! https://github.com/Significant-Gravitas/Auto-GPT/pull/4167

**Comment by Boostrix** [comment]
unless I am mistaken, this should not be closed or "staled" at all, I believe this remains relevant or has something changed over the course of the last couple of months that I missing entirely ?
